### ISSUE TITLE
Crash on remove_wsi_socket_from_fds (using libuv)

### ISSUE THREAD

**Issue by javiserrano** [warmcat/libwebsockets#1245]
Hello,

We are using libwebsockets (branch `v2.4-stable`) with libuv on a native library for iOS and Android and, in both platforms, we are getting crash reports like this one in production:

```
0  remove_wsi_socket_from_fds (pollfd.c:269)
1  remove_wsi_socket_from_fds (pollfd.c:259)
2  lws_close_free_wsi (libwebsockets.c:707)
3  lws_service_fd_tsi (service.c:1684)
4  lws_service_fd (service.c:1700)
5  lws_io_cb (libuv.c:98)
6  uv__io_poll (kqueue.c:0)
7  uv_run (core.c:368)
```

Unfortunately, we have been unable to reproduce the crash by ourselves, so the information we have is basically that stack trace. I suspect that the problem is related to a race condition while connections are being closed.

In order to try to repruduce it, I have duplicated the call to `lws_close_free_wsi` here, trying to simulate something like a connection closed by the server at the same time as the client, the reception of queued data while the connection is being closed or something similar.
https://github.com/warmcat/libwebsockets/blob/v2.4-stable/lib/service.c#L1684

Duplicating that function call, I get a crash with the same stack trace.

After more investigation, I have noticied that the disconnection when libuv is being used is asynchronous:
https://github.com/warmcat/libwebsockets/blob/v2.4-stable/lib/libwebsockets.c#L803

So maybe the same function is being called a second time before the handle is finally closed, causing the crash.

I have also noticed that `socket_is_permanently_unusable` is set to `1` few lines above, and `wsi->state` is set to `LWSS_DEAD_SOCKET` more lines above, here:
https://github.com/warmcat/libwebsockets/blob/v2.4-stable/lib/libwebsockets.c#L716

Both values seem to be related to the crash that I'm getting when I call `lws_close_free_wsi` twice, so maybe they are related to the crash that we get in production too. When I have debugged the second call to `lws_close_free_wsi` (the one that crashes in my tests), I have seen this interesting code:

https://github.com/warmcat/libwebsockets/blob/v2.4-stable/lib/libwebsockets.c#L448

```	
        if (wsi->socket_is_permanently_unusable ||
	    reason == LWS_CLOSE_STATUS_NOSTATUS_CONTEXT_DESTROY ||
	    wsi->state == LWSS_SHUTDOWN)
		goto just_kill_connection;

	wsi->state_pre_close = wsi->state;

	switch (wsi->state_pre_close) {
	case LWSS_DEAD_SOCKET:
		return;
``` 

It seems that it's intended to just return if `wsi->state` has the value `LWSS_DEAD_SOCKET`. However, since `socket_is_permanently_unusable` is also set, the code skips the second check and goes to `just_kill_connection`. If the funtion should really return due to the `LWSS_DEAD_SOCKET` status,  something like this above that code could fix the problem:

```        
        if (wsi->state == LWSS_DEAD_SOCKET)
        {
            wsi->state_pre_close = wsi->state;
            return;
        }
```
        
It seems to work fine (with that change, I can call `lws_close_free_wsi` many times in a row without getting a crash) but I'm not sure if that's a proper solution because my understanding or the internals of libwebsockets and its usage of libuv is very limited.

Do you think that the change would make sense and that it would be possible to get that crash without modifying the code, just due to a race condition during the asynchronous close of the libuv handle?

It would be great to know your opinion to see if I'm on the right track, or my guesses about the possible cause of the crash that we are getting in production are totally wrong.

Thank you very much.

**Comment by lws-team** [comment]
Thanks for the nice debugging and explanation.

> In order to try to reproduce it, I have duplicated the call to lws_close_free_wsi here

That's actually a cool idea... I added this on master if it's a debug build and libuv selected, it will always do the second close in order to confirm there's no weakness about it introduced.  The close "caller" pointer is set to "close_and_handled uv repeat test" on the second one so it's clear what happened.

On master with http/2 this doesn't make any problem even with valgrind... master has been refactored a lot compared to v2.4 though.  I applied your change anyway and confirmed no problem without h2.

On v2.4-stable I applied your change and again the double-close test if running uv and built for debug.  Hopefully that should close the window for double-closes making trouble on uv permanently.
